### DISCUSSION TITLE
Extractor limited to substates implementing given trait

### DISCUSSION THREAD

**Discussion by przucidlo** [tokio-rs/axum#2180] [Q&A]
Hey, I want to create an extractor that can be only applied to certain substates that implement a given trait. 

I've made it as far to make checks pass on handler arguments, but applying the handler to router yields an error that root state does not implement that trait. I can't come up with a decent idea how to make the compiler happy.

I'll skip my miserable attempts and present you with ideal scenario:

```rust
pub trait SomeTrait {
    fn some_function() -> ();
}

pub struct FancyExtractor(pub SomeStruct);

#[async_trait]
impl<S> FromRequestParts<S> for FancyExtractor
where
    S: Send + Sync + SomeTrait,
{
    type Rejection = StatusCode;

    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection> {
        s.some_function();

        Err(StatusCode::INTERNAL_SERVER_ERROR)
    }
}
```

I tried to play around with ```FromRef<State>``` from State extractor but I hit a wall, when I used a concrete type instead of generics on it.

Help would be highly appreciated, thanks!

**Comment by jplatte** [comment]
I don't think what you're trying to do is possible. What do you expect to happen if there are multiple types that implement both `SomeTrait` and `FromRef<RootState>`? How would the compiler know which `FromRef` / `SomeTrait` implementation is supposed to be used?
